Fixed bad shift operation warning in deflatePrime.

nmoinvaz opened the pull request:
```
  CID 293475 (#2-4 of 4): Bad bit shift operation (BAD_SHIFT)
  In expression 1UL << put, left shifting by more than 63 bits has undefined behavior.
```

nmoinvaz committed:
Fixed bad shift operation warning in deflatePrime.
Check that bits value is not greater than bits allowed by value type.

  CID 293475 (#2-4 of 4): Bad bit shift operation (BAD_SHIFT)
  In expression 1UL << put, left shifting by more than 63 bits has undefined behavior.

mtl1979 commented:
If ```s->bi_valid == 0```, then it should just copy and zero-extend the value instead of trying to mask and shift... The original calculation did overflow for obvious reason.

mtl1979 commented on deflate.c:
```diff
@@ -590,7 +590,7 @@ int32_t ZEXPORT PREFIX(deflatePrime)(PREFIX3(stream) *strm, int32_t bits, int32_
         return Z_BUF_ERROR;
     do {
-        put = BIT_BUF_SIZE - s->bi_valid;
+        put = (BIT_BUF_SIZE-1) - s->bi_valid;
```
This essentially reduces the bit buffer size by one bit, which is a bad thing...

nmoinvaz replied:
Fixed with your changes.

mtl1979 commented on deflate.c:
```diff
@@ -590,7 +590,7 @@ int32_t ZEXPORT PREFIX(deflatePrime)(PREFIX3(stream) *strm, int32_t bits, int32_
+        put = (BIT_BUF_SIZE-1) - s->bi_valid;
         if (put > bits)
             put = bits;
         s->bi_buf |= (((uint64_t)value & ((UINT64_C(1) << put) - 1)) << s->bi_valid);
```
```
if (s->bi_valid == 0)
    s->bi_buf = (uint64_t)value;
else
    s->bi_buf |= (((uint64_t)value & ((UINT64_C(1) << put) - 1)) << s->bi_valid);
```

nmoinvaz replied:
Ok it is fixed now. Thanks.

nmoinvaz commented:
Changes made.

mtl1979 commented on deflate.c:
```diff
@@ -593,7 +593,10 @@ int32_t ZEXPORT PREFIX(deflatePrime)(PREFIX3(stream) *strm, int32_t bits, int32_
             put = bits;
-        s->bi_buf |= (((uint64_t)value & ((UINT64_C(1) << put) - 1)) << s->bi_valid);
+        if (s->bi_valid == 0)
+            s->bi_buf = (uint64_t)value;
```
Ignoring ```bits``` is safe here because below we essentially set ```s->bi_valid```  to value of ```put``` (actually ```0 + put```) and the test ```put > bits``` above is outside the ```if``` block... Any extra bits will be shifted away when adding more bits to ```s->bi_buf```.

nmoinvaz replied:
@mtl1979, was there something you wanted me to do or were you just commenting?

mtl1979 replied:
@nmoinvaz I was just wondering if it needs a comment to explain why it ignores ```bits``` when ```s->bi_valid``` is 0.

nmoinvaz replied:
I don't really think so personally.

Dead2 commented:
Looks good to me, but I wish we had a test for this. Is that a possibility here?

nmoinvaz committed:
Improve deflatePrime test by wrapping gzip header and footer around deflate stream.

nmoinvaz committed:
Fixed integer casting and signed comparison warning in test_gzio.

    example.c(199,57): warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data
    example.c:207:32: warning: comparison between signed and unsigned integer expressions

nmoinvaz commented:
I have made some improvements to `deflatePrime` testing. There is not a possibility for `bits` param being 64 because `value` param is 32-bit so the problem this issue solves would never be hit. But it had been fixed to silence coverity.

Dead2 commented:
This currently fails all checks. Unfortunately I don't know why, because github actions says
`We are currently unable to download the log. Please try again later.`

mtl1979 commented:
```/home/runner/work/zlib-ng/zlib-ng/test/example.c:1055:5: error: too few arguments to function ‘test_deflate_prime’```

mtl1979 commented on test/example.c:
```diff
@@ -805,50 +806,96 @@ void test_deflate_pending(unsigned char *compr, size_t comprLen) {
+ * Test deflatePrime() wrapping gzip around deflate stream
  */
-void test_deflate_prime(unsigned char *compr, size_t comprLen) {
+void test_deflate_prime(unsigned char *compr, size_t comprLen, unsigned char *uncompr, size_t uncomprLen) {
```
This is bad... Changes in function signature break build.

nmoinvaz replied:
Fixed thanks.

Dead2 commented:
This should also fix the newly found coverity issue at line 599 `value >>= put;` (It basically detects this issue in four variants now)

Dead2 commented:
When you fix the current compile error, please merge the (now 2, would-be 3) test-commits into one. 😄

nmoinvaz commented:
They are not test commits, I have broken them out by the logic changes. However, I have merged the first two into one commit to reduce the number of commits. Hopefully that looks good to you.

Dead2 commented:
My mistake then, I only reviewed the whole thing and then found out there were multiple commits, and assumed the last one was a fix for the next-to-last one. Sorry.

Btw, is Github working really crappy for you lately? Click on "commits" or "files changed" and it never finishes loading until you refresh, same when you want to return to "conversation". Has made it a royal pain to review things the past couple of weeks. Might be something on my end, but not sure what would cause that particular problem, kind of like ajax-requests just get lost on the way.

mtl1979 commented:
@Dead2 GitHub status page is showing a lot of orange and red for last few days...
